cmd/config.go

```diff
@@ -0,0 +1,217 @@
+	WSListeners     []*backends.WebsocketListenerCfg `mapstructure:"ws-listeners"`
+	TLSClients      []netceptor.TLSClientConfig      `mapstructure:"tls-clients"`
+	TLSServer       []netceptor.TLSServerConfig      `mapstructure:"tls-servers"`
+	WorkCommands    []workceptor.CommandWorkerCfg    `mapstructure:"work-commands"`
+	WorkKubernetes  []*workceptor.KubeWorkerCfg      `mapstructure:"work-kubernetes"`
+	// WorkPython        []workceptor.WorkPythonCfg       `mapstructure:"work-python"`
+	WorkSigning       workceptor.SigningKeyPrivateCfg  `mapstructure:"work-signing"`
+	WorkVerification  workceptor.VerifyingKeyPublicCfg `mapstructure:"work-verification"`
+	IpRouters         []services.IpRouterCfg           `mapstructure:"ip-routers"`
+	TCPClients        []services.TCPProxyOutboundCfg   `mapstructure:"tcp-clients"`
+	TCPServers        []services.TCPProxyInboundCfg    `mapstructure:"tcp-servers"`
+	UDPClients        []services.TCPProxyInboundCfg    `mapstructure:"udp-clients"`
+	UDPServers        []services.TCPProxyInboundCfg    `mapstructure:"udp-servers"`
+	UnixSocketClients []services.UnixProxyOutboundCfg  `mapstructure:"unix-socket-clients"`
+	UnixSocketServers []services.UnixProxyInboundCfg   `mapstructure:"unix-socket-servers"`
+}
+
+type CertificatesConfig struct {
+	InitCA  certificates.InitCAConfig    `mapstructure:"cert-init"`
+	MakeReq []certificates.MakeReqConfig `mapstructure:"cert-makereq"`
+	SignReq []certificates.SignReqConfig `mapstructure:"cert-signreq"`
+}
+
+func PrintPhaseErrorMessage(configName string, phase string, err error) {
+	fmt.Printf("ERROR: %s for %s on %s phase\n", err, configName, phase)
+}
+
+func ParseConfigs(configFile string) (*ReceptorConfig, *CertificatesConfig, error) {
+	if configFile == "" {
+		fmt.Fprintln(os.Stderr, "Could not locate config file (default is $HOME/.receptor.yaml)")
+		os.Exit(1)
+	}
+	var receptorConfig ReceptorConfig
+	var certifcatesConfig CertificatesConfig
+	err := viper.Unmarshal(&receptorConfig)
+	if err != nil {
+		return nil, nil, err
+
+	}
+
+	err = viper.Unmarshal(&certifcatesConfig)
+	if err != nil {
+		return nil, nil, err
+
+	}
+	return &receptorConfig, &certifcatesConfig, nil
+}
+
+func isConfigEmpty(v reflect.Value) bool {
+	isEmpty := true
+	for i := 0; i < v.NumField(); i++ {
+		if reflect.Value.IsZero(v.Field(i)) {
+			continue
+		}
+		isEmpty = false
+	}
+	return isEmpty
+}
+
+func RunConfigV2(v reflect.Value) {
+	phases := []string{"Init", "Prepare", "Run"}
+
+	for _, phase := range phases {
+		for i := 0; i < v.NumField(); i++ {
+			if reflect.Value.IsZero(v.Field(i)) {
+				continue
+			}
+
+			switch v.Field(i).Kind() {
+			case reflect.Slice:
+				for j := 0; j < v.Field(i).Len(); j++ {
+					RunPhases(phase, v.Field(i).Index(j))
+				}
+			default:
+				RunPhases(phase, v.Field(i))
+			}
+
+		}
+	}
+}
+
+// RunPhases runs the appropriate function (Init, Prepare, Run) on a command
+func RunPhases(phase string, v reflect.Value) {
```

Comment:
Did we ever run that test to see if the phases need to be run in order?

Reply:
We did not. We decided to do performance improvements after, this includes removing the dependency on the receptor binary for work